File: carddrawer/src/test/java/com/meli/android/carddrawer/model/GradientTextViewTest.java
package com.meli.android.carddrawer.model;

import android.graphics.Canvas;
import android.text.TextPaint;

import com.meli.android.carddrawer.BasicRobolectricTest;
import com.meli.android.carddrawer.configuration.CardFontConfiguration;
import com.meli.android.carddrawer.configuration.DarkFontConfiguration;
import com.meli.android.carddrawer.configuration.FontType;

import org.junit.Test;
import org.junit.runner.RunWith;

import org.robolectric.RobolectricTestRunner;
import org.robolectric.util.ReflectionHelpers;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;

@RunWith(RobolectricTestRunner.class)
public class GradientTextViewTest extends BasicRobolectricTest {

    @Test
    public void init_loadsViews() {
        GradientTextView gradientTextView = new GradientTextView(getContext());

        assertNotNull(ReflectionHelpers.getField(gradientTextView, "fontType"));
        assertNotNull(ReflectionHelpers.getField(gradientTextView, "placeHolder"));
        assertNotNull(ReflectionHelpers.getField(gradientTextView, "fontColor"));
    }

    @Test
    public void onDraw_setConfigurationGradient() {
        GradientTextView gradientTextViewSpy = spy(new GradientTextView(getContext()));
        CardFontConfiguration configurator = mock(CardFontConfiguration.class);
        doReturn(configurator).when(gradientTextViewSpy).getConfiguration();
        doReturn("anotherCardNumber").when(gradientTextViewSpy).getText();
        ReflectionHelpers.setField(gradientTextViewSpy, "placeHolder", "cardNumber");

        Canvas canvas = mock(Canvas.class);

        gradientTextViewSpy.onDraw(canvas);

        verify(configurator).setGradient(any(TextPaint.class), anyInt(), anyInt());
        verify(configurator).setShadow(any(TextPaint.class));
    }

    @Test
    public void onDraw_doesntSetConfigurationGradient() {
        GradientTextView gradientTextViewSpy = spy(new GradientTextView(getContext()));
        CardFontConfiguration configurator = mock(CardFontConfiguration.class);
        doReturn(configurator).when(gradientTextViewSpy).getConfiguration();
        doReturn("cardNumber").when(gradientTextViewSpy).getText();
        ReflectionHelpers.setField(gradientTextViewSpy, "placeHolder", "cardNumber");

        Canvas canvas = mock(Canvas.class);

        gradientTextViewSpy.onDraw(canvas);

        verify(configurator, never()).setGradient(any(TextPaint.class), anyInt(), anyInt());
        verify(configurator, never()).setShadow(any(TextPaint.class));
    }

    @Test
    public void init_setsValues() {
        GradientTextView gradientTextView = new GradientTextView(getContext());
        gradientTextView.init(FontType.LIGHT_TYPE, "MM/YY", 2);

        assertEquals(FontType.LIGHT_TYPE, ReflectionHelpers.getField(gradientTextView, "fontType"));
        assertEquals("MM/YY", ReflectionHelpers.getField(gradientTextView, "placeHolder"));
        assertEquals(2, ReflectionHelpers.getField(gradientTextView, "fontColor"));
    }

    @Test
    public void getConfiguration_lalal() {
        GradientTextView gradientTextViewSpy = spy(new GradientTextView(getContext()));
        ReflectionHelpers.setField(gradientTextViewSpy, "fontType", FontType.DARK_TYPE);
        ReflectionHelpers.setField(gradientTextViewSpy, "fontColor", 2);

        assertTrue(gradientTextViewSpy.getConfiguration() instanceof DarkFontConfiguration);
    }

}

File: carddrawer/src/test/java/com/meli/android/carddrawer/model/GradientTextViewTest.kt
package com.meli.android.carddrawer.model

import android.graphics.Canvas
import com.meli.android.carddrawer.BasicRobolectricTest
import com.meli.android.carddrawer.configuration.CardFontConfiguration
import com.meli.android.carddrawer.configuration.FontType
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.util.ReflectionHelpers

@RunWith(RobolectricTestRunner::class)
class GradientTextViewTest : BasicRobolectricTest() {

    @Test
    fun `when init view then initialize fields with values`() {
        val gradientTextView = GradientTextView(context)
        Assert.assertNotNull(ReflectionHelpers.getField(gradientTextView, "fontType"))
        Assert.assertNotNull(ReflectionHelpers.getField(gradientTextView, "placeHolder"))
        Assert.assertNotNull(ReflectionHelpers.getField(gradientTextView, "fontColor"))
    }

    @Test
    fun `when call function onDraw with text 'anotherCardNumber' then call setShadow`() {
        val gradientTextView = mockk<GradientTextView>(relaxed = true)
        val configurator = mockk<CardFontConfiguration>(relaxed = true)
        val canvas = mockk<Canvas>(relaxed = true)
        every { gradientTextView.onDraw(canvas) } answers { callOriginal() }
        every { gradientTextView.configuration } returns configurator
        every { gradientTextView.text } returns "anotherCardNumber"
        ReflectionHelpers.setField(gradientTextView, "placeHolder", "cardNumber")
        gradientTextView.onDraw(canvas)
        verify {
            configurator.setShadow(
                any()
            )
        }
    }

    @Test
    fun `when call function onDraw with text 'cardNumber' then shouldn't call setShadow`() {
        val gradientTextView = mockk<GradientTextView>(relaxed = true)
        val configurator = mockk<CardFontConfiguration>(relaxed = true)
        val canvas = mockk<Canvas>(relaxed = true)
        every { gradientTextView.onDraw(canvas) } answers { callOriginal() }
        every { gradientTextView.configuration } returns configurator
        every { gradientTextView.text } returns "cardNumber"
        ReflectionHelpers.setField(gradientTextView, "placeHolder", "cardNumber")
        gradientTextView.onDraw(canvas)
        verify(inverse = true) {
            configurator.setShadow(
                any()
            )
        }
    }

    @Test
    fun `when call function init then initialize fields with values`() {
        val gradientTextView = GradientTextView(context)
        gradientTextView.init(FontType.LIGHT_TYPE, "MM/YY", 2)
        Assert.assertEquals(
            FontType.LIGHT_TYPE,
            ReflectionHelpers.getField(gradientTextView, "fontType")
        )
        Assert.assertEquals("MM/YY", ReflectionHelpers.getField(gradientTextView, "placeHolder"))
        Assert.assertEquals(
            2,
            ReflectionHelpers.getField<Any>(gradientTextView, "fontColor") as Int
        )
    }
}